src/transformers/models/isaac/modular_isaac.py

```diff
@@ -0,0 +1,2109 @@
+            entries gives each document's token count and enables block-diagonal attention masking for packed batches.
+        max_seqlen (`int`, *optional*):
+            Maximum document length referenced by `cu_seqlens`. Passed to FlashAttention so it can size temporary
+            buffers for packed variable-length attention.
+        """
+        attention_mask = ensure_document_attention_mask(
+            attention_mask,
+            cu_seqlens,
+            hidden_states.size(1),
+            hidden_states.dtype,
+            hidden_states.device,
+            return_mask_function=False,
+        )
+
+        # Run attention directly so variable-length metadata reaches FlashAttention.
+        residual = hidden_states
+        hidden_states = self.layer_norm1(hidden_states)
+        attn_outputs = self.self_attn(
+            hidden_states,
+            attention_mask=attention_mask,
+            cu_seqlens=cu_seqlens,
+            max_seqlen=max_seqlen,
+            output_attentions=output_attentions,
+            **kwargs,
+        )
+        if isinstance(attn_outputs, tuple):
+            attn_output, attn_weights = attn_outputs
+        else:
+            attn_output, attn_weights = attn_outputs, None
+        hidden_states = residual + attn_output
+
+        residual = hidden_states
+        hidden_states = self.layer_norm2(hidden_states)
+        hidden_states = self.mlp(hidden_states)
+        hidden_states = residual + hidden_states
+
+        if output_attentions:
+            return hidden_states, attn_weights
+        return hidden_states
+
+
+class IsaacVisionEncoder(Siglip2Encoder):
+    """Encoder using Isaac encoder layers with variable-length attention support."""
+
+    def __init__(self, config: IsaacVisionConfig):
+        super().__init__(config)
+        self.layers = nn.ModuleList([IsaacVisionEncoderLayer(config) for _ in range(config.num_hidden_layers)])
+
+    @can_return_tuple
+    def forward(
+        self,
+        inputs_embeds,
+        attention_mask: Optional[torch.Tensor] = None,
+        cu_seqlens: Optional[torch.Tensor] = None,
+        max_seqlen: Optional[int] = None,
+        output_attentions: Optional[bool] = None,
+        output_hidden_states: Optional[bool] = None,
+        return_dict: Optional[bool] = None,
+        **kwargs: Unpack[TransformersKwargs],
+    ):
+        attention_mask = ensure_document_attention_mask(
+            attention_mask,
+            cu_seqlens,
+            inputs_embeds.size(1),
+            inputs_embeds.dtype,
+            inputs_embeds.device,
+            return_mask_function=False,
+        )
+
+        hidden_states = inputs_embeds
+        kwargs.update(
+            {
+                "max_seqlen": max_seqlen,
+                "cu_seqlens": cu_seqlens,
+                "output_attentions": output_attentions,
+                "output_hidden_states": output_hidden_states,
+                "return_dict": return_dict,
+            }
+        )
```

Comment:
why do you need to recreate them? just leave them as kwarg. For document masking read from kwargs